vinumc: expose eval context to external functions

artP2 opened the pull request:
This allow the external library to get arguments from the function call by index and to get any name available on the call scope

artP2 committed:
vinumc: expose eval context to external functions

allows creation of functions that enable external functions to get
args and symbols from internal ast and perform actions on it

artP2 committed:
vinumc: use compiler ctx on eval functions

We dont need to split the context fields to pass it to the evaluation
functions.
This also simplifies exposing the context to external funcions.

artP2 committed:
vinumc/eval: Add a function to find symbols on scope

We want tocreate a functoin to be used by externl libraries that wil
return a text of a symbol. This function will use the
eval_find_symbol_on_scopes.

artP2 committed:
vinumc: extern_library add ctx_eval_symbol

allows the extern library to get a symbol value by its name

JeanJPNM commented on subprojects/vinumc/include/vinumc/extern_library.h:
```diff
@@ -24,5 +24,7 @@ struct loaded_lib {
 
 struct return_value ctx_get_text(call_ctx ctx);
+struct return_value ctx_call(call_ctx ctx);
+struct return_value ctx_get_index(call_ctx ctx, int index);
```
No way of knowing the total amount of arguments.
No way of accessing individual parameters by name.

```c
int ctx_get_arg_count(call_ctx ctx);
call_ctx_or_null get_arg_by_name(call_ctx ctx);
```

artP2 replied:
this part is going to be moved to another PR to speedup development.

JeanJPNM commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,21 @@
+struct return_value ctx_get_index(call_ctx ctx, int index) {
+	if (ctx->status == true) {
+		return (struct return_value){ .ptr = VUT_VEC_POP(&ctx->args), .status = true };
+	}
```
No proper range checking.

JeanJPNM commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,21 @@
+struct return_value ctx_call(call_ctx ctx) {
+	if (ctx->status == true) {
+		return (struct return_value){ .ptr = VUT_VEC_POP(&ctx->names), .status = true };
+	}
```
Needs to check if `ctx->names` is not empty.

artP2 replied:
this part is going to be moved to another PR to speedup development.

JeanJPNM commented on subprojects/vutils/include/vutils/vec.h:
```diff
@@ -67,13 +67,8 @@
-		}                                                                                  \
-	} while (0)
+/// Returns the last element and decrements the length of the vector by one.
+#define VUT_VEC_POP(arr) ((arr)->len > 0 ? (arr)->base[--(arr)->len] : (arr)->base[0])
```
Returns (possibly) garbage values when the vector is empty, may be better to read the last element before popping. You could create a `VUT_VEC_LAST` macro.

artP2 replied:
this part is going to be moved to another PR to speedup development.

JeanJPNM commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,20 @@
+	struct v_str_vec args;
+	struct v_str_vec names;
+	struct arg_vec args_req;
+	struct name_vec names_req;
```
Missing doc comments for added fields

Grillo-0 commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,20 @@
 #define __V_LIB_H__
 
+#include <stdbool.h>
+#include <vutils/vec.h>
+
+struct v_str_vec VUT_VEC_DEF(char *);
+struct arg_vec VUT_VEC_DEF(int);
+struct name_vec VUT_VEC_DEF(char *);
```
I don't like using vutils in the header. This adds an obligatory dependency to the developer of the external lib.

To help things, you could create a vec macro that gives you the base pointer and length to put it in the ctx.

Something like that:
```c
int* ptr;
size_t len;
VUT_VEC_GET_BASE_ARRAY(&vec, ptr, len)
```

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -257,8 +257,38 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
+			}
+
+			cctx.status = true;
+			call_return = symbol_info->as.func(&cctx);
```
We can't call the the function twice on a single vinum call.

What if the external function is not [idempotent](https://en.wikipedia.org/wiki/Idempotence)? For example, a function that adds a single entry to a database: If we call it on vinum code, two entries could be added instead of just a single one.

Is it possible to pre-compute the args before calling the external function?

artP2 replied:
> Is it possible to pre-compute the args before calling the external function?

This is the biggest change, so I'll focus on it first. Precomputing the arguments is possible, which is how I did it before, but for example, to search for a name in the namespace, we would have to provide the scope and the function to search for symbols.
The current method allows for flexibility, without having to provide the internal functions to the external library. To ensure it's idempotent, I thought about requiring the external library to check if ctx.status is true before performing the operations.
I would like to discuss this thoroughly before making any changes, since it alters the entire implementation.
In any case, suggestions are welcome.

artP2 commented:
Rebased on #117

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -93,20 +110,20 @@ static int find_scope_child_by_node(const struct eval_ctx_scopes_t *scopes, size
 #define RESOLVE_FUNC_SIGNATURE(func_name)                                                          \
-	static void func_name(struct eval_ctx *ctx, struct ast *ast, size_t curr_scope_id,         \
-			      size_t ast_node_id)
+	static void func_name(struct compiler_ctx *cctx, size_t curr_scope_id, size_t ast_node_id)
```
Split this on its own commit. Also, while at it, do this:
```suggestion
	static void func_name(struct compiler_ctx *cctx, size_t curr_scope_id, ast_node_id_t ast_node)
```

With the new AST functions we hardly need to access the nodes directly, so we don't need to put an `_id` suffix to it

Do this for all signatures.

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -276,21 +296,29 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
 		// so any nested call will be resolved normally
-		struct vut_str tmp_out = vut_str_init(ctx->allocator);
-		do_calls(ctx, ast, &tmp_out, args, flags);
+		struct vut_str tmp_out = vut_str_init(cctx->alloc);
```
Maintain the use of the eval allocator

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -76,6 +76,23 @@ static struct namespace_entry *find_symbol_on_scopes(const struct eval_ctx_scope
 }
 
+static ssize_t find_scope_by_ast_node_id(const struct eval_ctx_scopes_t *scope_array,
+					 size_t ast_node_id) {
```
```suggestion
static ssize_t find_scope_by_ast_node(const struct eval_ctx_scopes_t *scope_array,
					 ast_node_id_t ast_node) {
```

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -380,18 +406,19 @@ void unload_libs(struct loaded_lib *loaded_libs, struct vut_allocator *alloc) {
 
-	struct vut_str str_out = vut_str_init(ctx->allocator);
-	do_calls(ctx, ast, &str_out, 0, REDUCE_BLANKS);
+	struct vut_str str_out = vut_str_init(cctx->alloc);
```
Same here, use the eval allocator

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -380,18 +406,19 @@ void unload_libs(struct loaded_lib *loaded_libs, struct vut_allocator *alloc) {
+	do_calls(cctx, &str_out, 0, REDUCE_BLANKS);
 
-	unload_libs(loaded_libs, ctx->allocator);
+	unload_libs(loaded_libs, cctx->alloc);
```
Same here, use the eval allocator

Grillo-0 commented on subprojects/vinumc/v_lib.c:
```diff
@@ -1,6 +1,27 @@
 	return (struct return_value){ .ptr = ctx->text, .status = true };
 }
+
+struct return_value ctx_get_symbol_by_name(call_ctx ctx, const char *name) {
```
```suggestion
struct return_value ctx_get_symbol_text_by_name(call_ctx ctx, const char *name) {
```

Grillo-0 reviewed (changes requested):
Improve the commit msg eon the second commit

artP2 committed:
vinumc: rename ast_node_id to ast_node

With the new ast functions we dont need to have a reference to a node to
use it, we only need the id, so we can simplify it's name.

Grillo-0 commented on subprojects/vinumc/tests/library_test.c:
```diff
@@ -41,6 +41,14 @@ void test_call_no_args(struct vunit_test_ctx *ctx) {
+
+	VUNIT_ASSERT_STREQ(ctx, out, "Ipsum");
+}
+
```
Add a test for

`[author_last_name [author: Lorem Ipsum]]`

artP2 commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,22 @@
+	}
+
+	return (struct return_value){};
+}
```
evaluate symbol text before

artP2 commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,22 @@
 	return (struct return_value){ .ptr = ctx->text, .status = true };
 }
+
+struct return_value ctx_get_symbol_text_by_name(call_ctx ctx, const char *name) {
```
rename to ctx_eval_symbol

artP2 commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,22 @@
 struct return_value ctx_get_text(call_ctx ctx) {
```
rename to ctx_get_arg

artP2 commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,16 @@
+#include "libvinumc.h"
+
 struct _call_ctx {
 	char *text;
```
rename to arg_text ?

artP2 commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,16 @@
 	char *text;
+	struct compiler_ctx *compiler_ctx;
+	size_t scope_id;
+	size_t ast_node_id;
```
rename to ast_node

artP2 commented on subprojects/.wraplock:
remove

artP2 committed:
vinumc: rename ctx_get_text to ctx_get_arg

better fit the usage of the function

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -292,14 +303,22 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
-		struct _call_ctx cctx = { .text = tmp_text };
-		struct return_value call_return = symbol_info->as.func(&cctx);
+		struct _call_ctx call_ctx = {
+			.arg_text = vut_str_move_to_cstr(&tmp_out),
```
You still need to free this.

Grillo-0 commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,16 @@
+	struct compiler_ctx *compiler_ctx;
+	size_t scope_id;
+	size_t ast_node;
+	bool status;
```
Why do wee need this `status`?

Grillo-0 commented on subprojects/vinumc/eval.c:
```diff
@@ -88,6 +88,12 @@ static int find_scope_by_ast_node(const struct eval_ctx_scopes_t *scope_array,
 	return -1;
 }
 
+struct namespace_entry *eval_find_symbol_on_scopes(const struct eval_ctx *ctx, size_t curr_scope,
```
Describe why you need this function in the commit description

Grillo-0 commented on subprojects/vinumc/v_lib.c:
```diff
@@ -4,3 +4,16 @@
+	return (struct return_value){
+		.ptr = vut_str_move_to_cstr(&text),
+		.status = true,
+		.free = false,
```
This should be `true`.

Grillo-0 commented on subprojects/vinumc/v_lib.h:
```diff
@@ -1,8 +1,16 @@
+
 struct _call_ctx {
-	char *text;
+	char *arg_text;
```
Move this arg_text change to the commit where you change the function name

artP2 commented:
bypassed cause @Grillo-0 did the remaining changes and I cannot aprove my own PR.
